### PR TITLE
Setgridfield

### PR DESCRIPTION
**Pull request by teizenman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/24452

## Notes
Got approval from both @idovandijk and @altmannyarden that it works for them.

### CODE REVIEW

File: Packs/CommonScripts/Scripts/SetGridField/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 ## [Unreleased]
--
+- Fixed an issue with a dictionary element.
+- Added support for lists of values.
+- Added support for unpacking nested elements.
+- The argument **keys** is no longer mandatory, all the keys will be taken by default.
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/CHANGELOG.md:6]: No need to have both rn types

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.yml [hunk 2]
```diff
@@ -30,18 +30,30 @@ args:
...
   isArray: false
   name: sort_by
   required: false
   secret: false
+- auto: PREDEFINED
+  default: false
+  defaultValue: 'true'
+  description: Set to 'false' to unpack nested elements.
+  isArray: false
+  name: skip_nested_elements
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.yml:51]: unpack vs skip, let's make it more consistent

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 3]
```diff
@@ -135,10 +214,10 @@ def build_grid(context_path: str, keys: List[str], columns: List[str]) -> pd.Dat
...
 
 def build_grid_command(grid_id: str, context_path: str, keys: List[str], columns: List[str], overwrite: bool,
-                       sort_by: str) \
+                       sort_by: str, skip_nested_elements: bool) \
         -> List[Dict[Any, Any]]:
-    """ Build Grid in one of the 2 options:
-            1. Context_path contain list, e.g. [{'a': 1, 'b': 2}, {'a': 1, 'b': 2}]
+    """ Build Grid in one of the 3 options:
+            1. Context_path contain list of dicts, e.g. [{'a': 1, 'b': 2}, {'a': 1, 'b': 2}]
             2. Context_path contain dict (key value pairs), e.g. {'a': 1, 'b': 2}
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:221]: where is 3?

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 4]
```diff
@@ -72,33 +115,61 @@ def get_current_table(grid_id: str) -> List[Dict[Any, Any]]:
...
+            if not isinstance(item, (str, int, float, bool)):
+                raise ValueError(exception_msg)
 
-def build_grid(context_path: str, keys: List[str], columns: List[str]) -> pd.DataFrame:
+    return data_type
+
+
+def build_grid(context_path: str, keys: List[str], columns: List[str], skip_nested_elements: bool) -> pd.DataFrame:
     """ Build new DateFrame from current context retrieved by DT.
         There is 3 cases:
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:174]:
```suggestion
        There are 3 cases:
```

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 5]
```diff
@@ -110,23 +181,31 @@ def build_grid(context_path: str, keys: List[str], columns: List[str]) -> pd.Dat
...
     # Building new Grid
-    if len(entry_context_data) > 1:
-        # Handle entry context list option
+    if not skip_nested_elements:
+        table = pd.DataFrame(unpack_all_data_from_dict(entry_context_data, keys, columns))
+        table.rename(columns=dict(zip(table.columns, columns)), inplace=True)
+    elif data_type == 'list':
+        # Handle entry context as list of value
+        table = pd.DataFrame(entry_context_data)
+        table.rename(columns=dict(zip(table.columns, columns)), inplace=True)
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:200]: don't you want to unpack here too? or is it not the usecase?

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 6]
```diff
@@ -72,33 +115,61 @@ def get_current_table(grid_id: str) -> List[Dict[Any, Any]]:
...
-    if not isinstance(entry_context, list):
+    exception_msg = "Not valid entry context path - dict[Any,Any]"
+    if not skip_nested_elements:
+        if not isinstance(entry_context, dict):
+            raise ValueError(exception_msg)
+        else:
+            return
+
+    exception_msg = "Not valid entry context path - dict[str,str] or list[dict[str,str]] or List[str/bool/int/float]"
+    if not isinstance(entry_context, (list, dict)):
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:141]: didn't get the error msg

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 7]
```diff
@@ -72,33 +115,61 @@ def get_current_table(grid_id: str) -> List[Dict[Any, Any]]:
...
+        return data_type
+
+    has_seen_dict = False
     for item in entry_context:
         if not isinstance(item, dict):
-            raise ValueError(exception_msg)
+            if not has_seen_dict:
+                break
+            else:
+                raise ValueError(exception_msg)
```
Comment (ronykoz) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:155]: why is that problematic, we can have dict in the table

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py (hunk 5 above)
Comment (teizenman) [reply Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:200]: This is not the use-case here, this is intended to a dict with primitive types in the values only.
Updated the comment.

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py (hunk 7 above)
Comment (teizenman) [reply Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:155]: In the mode of `unpack_nested_elements` set to False, the context object should be either
- dict with primitive values only.
- list with primitive values only.
- list of dicts with primitive values only.

Therefore, this line catches the situation where I have a mix of primitive and non-primitive values in the list.

**Comment by teizenman** [comment]
@kirbles19 Can you please go over the following files:
- https://github.com/demisto/content/blob/setgridlist/Packs/CommonScripts/ReleaseNotes/1_1_0.md
- Arguments `unpack_nested_elements` and `keys` in: https://github.com/demisto/content/blob/setgridlist/Packs/CommonScripts/Scripts/SetGridField/SetGridField.yml

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py [hunk 8]
```diff
@@ -72,61 +115,102 @@ def get_current_table(grid_id: str) -> List[Dict[Any, Any]]:
...
 
+        has_seen_dict = True
         for key, value in item.items():
             if key in keys:
                 if not isinstance(value, (str, int, float, bool)):
```
Comment (reutshal) [inline Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:165]: Didn't you say that the dict inside a list should only contain str values?

File: Packs/CommonScripts/pack_metadata.json [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
-  "email": "",
-  "created": "2020-05-14T08:33:25Z",
-  "categories": [],
-  "tags": [],
-  "useCases": [],
-  "keywords": []
+    "name": "Common Scripts",
+    "description": "Frequently used scripts pack.",
+    "support": "xsoar",
+    "currentVersion": "1.1.0",
```
Comment (reutshal) [inline Packs/CommonScripts/pack_metadata.json:5]:
```suggestion
    "currentVersion": "1.0.1",
```

File: Packs/CommonScripts/ReleaseNotes/1_1_0.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (reutshal) [inline Packs/CommonScripts/ReleaseNotes/1_1_0.md:1]: If you are changing the version in the pack metadata you should probably change the name of this file too

File: Packs/CommonScripts/Scripts/SetGridField/SetGridField.py (hunk 8 above)
Comment (teizenman) [reply Packs/CommonScripts/Scripts/SetGridField/SetGridField.py:165]: Only primitive values

**Comment by kirbles19** [comment]
@teizenman Done.

**Pull request merged**